### PR TITLE
Sdl2

### PR DESCRIPTION
**Pull request by fabiangreffrath**

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -26,9 +26,8 @@ CFLAGS += -Wreturn-type
 CFLAGS += -Wwrite-strings
 CFLAGS += -Wcast-align
 
-ifdef GPL
-    CFLAGS += -DUSE_GPL
-endif
+CFLAGS += -DUSE_GPL
+CFLAGS += -fno-toplevel-reorder
```
Comment (turol) [inline Makefile:42]: Why?

File: id_in.cpp [hunk 2]
```diff
@@ -37,10 +37,10 @@ boolean forcegrabmouse;
 
 
 // 	Global variables
-volatile boolean    Keyboard[SDLK_LAST];
+std::map<ScanCode, boolean>	Keyboard;
```
Comment (turol) [inline id_in.cpp]: Prefer `unordered_map`

File: id_vl.cpp [hunk 3]
```diff
@@ -96,43 +100,91 @@ void	VL_Shutdown (void)
...
 
-    if(screenBits == -1)
+    window = SDL_CreateWindow(title,
+                              SDL_WINDOWPOS_CENTERED, SDL_WINDOWPOS_CENTERED,
+                              screenWidth, screenHeight,
+                              SDL_WINDOW_ALLOW_HIGHDPI | SDL_WINDOW_RESIZABLE |
+                              (fullscreen ? SDL_WINDOW_FULLSCREEN_DESKTOP : 0));
+    if (!window)
+    {
+        printf("Unable to create %ix%i window: %s\n",
```
Comment (turol) [inline id_vl.cpp]: `%u` for unsigned

File: Makefile (hunk 1 above)
Comment (fabiangreffrath) [reply Makefile:42]: Funny, I just had a conversation with @jwrdegoede about this. In contrast to Doom, which has all its possible game states enumerated  and stored in a single huge `info.c:states[]` array, the Wolf3d engine has them as global variables in `wl_act2.cpp`. The savegame code relies on the order of these variables remaining constant in the executable and the difference of these variables' addresses relative to a certain reference as well. This should be a valid assumption for a given executable, but might change when the code is compiled again - turning the savegames saved wit the prior compilate useless. This compiler flag prevents the reordering of the global variables and thus keeps savegames functional across engine rebuilds.

File: id_in.cpp (hunk 2 above)
Comment (fabiangreffrath) [reply id_in.cpp]: Okay, thanks!

How about here?
https://github.com/fabiangreffrath/wolf4sdl/pull/14/files#diff-03cbd5ce95e2076a4e78326aaca912fc

File: Makefile (hunk 1 above)
Comment (turol) [reply Makefile:42]: Put that comment in the Makefile. Also file an issue about fixing that, it's really not a safe assumption even with that flag.

File: id_in.cpp (hunk 2 above)
Comment (turol) [reply id_in.cpp]: Yes, `unordered_map` is always preferred when you don't specifically need in-order iteration over the container. Normal `map` is a balanced tree and can waste a lot of effort walking the tree.

File: Makefile (hunk 1 above)
Comment (fabiangreffrath) [reply Makefile:42]: https://github.com/fabiangreffrath/wolf4sdl/issues/15

**Pull request merged**